File: algorithms/closest_pair.cpp
/* 
Adaptação do código de Edson Alves, disponível em:
https://github.com/edsomjr/TEP/blob/master/Geometria_Computacional/slides/SL-2/closest.cpp
*/ 

#include <bits/stdc++.h>

using namespace std;
using point = pair<int, int>;

#define x first
#define y second

double get_distance(const point& P, const point& Q){
    return hypot(P.x - Q.x, P.y - Q.y);
}

pair<point, point> get_closest_pair(vector<point>& ps){
    size_t N = ps.size();
    sort(ps.begin(), ps.end());

    // Assume que N > 1
    auto distance = get_distance(ps[0], ps[1]);
    auto closest = make_pair(ps[0], ps[1]);

    set<point> S;
    S.insert(point(ps[0].y, ps[0].x));
    S.insert(point(ps[1].y, ps[1].x));

    for (int i = 2; i < N; ++i){
        auto P = ps[i];
        auto it = S.lower_bound(point(P.y - distance, 0));

        while (it != S.end()){
            auto Q = point(it->second, it->first);

            if (Q.x < P.x - distance){
                it = S.erase(it);
                continue;
            }

            if (Q.y > P.y + distance)
                break;

            auto curr_distance = get_distance(P, Q);

            if (curr_distance < distance){
                distance = curr_distance;
                closest = make_pair(P, Q);
            }

            ++it;
        }

        S.insert(point(P.y, P.x));
    }

    return closest;
}

int main(){
    vector<point> P {
        {2, 4},
        {5, 3},
        {8, 1},
        {3, 6},
        {1, 1},
        {4, 4},
        {8, 5},
        {6, 2},
        {1, 6},
        {7, 4},
        {8, 6},
        {3, 2},
        {6, 5},
        {-1, 4}
    };

    auto closest_pair = get_closest_pair(P);

    printf("A: (%d, %d), B: (%d, %d)\n",
           closest_pair.first.first, closest_pair.first.second,
           closest_pair.second.first, closest_pair.second.second
    );

    return 0;
}

File: algorithms/closest_pair.py
import math
from collections import namedtuple
from bisect import bisect_left


Point = namedtuple('Point', 'x y')
Pair = namedtuple('Pair', 'A B')


def get_distance(P, Q):
    return math.hypot(P.x - Q.x, P.y - Q.y)


def get_closest_pair(points):
    N = len(points)
    points.sort()

    distance = get_distance(points[0], points[1])
    closest = Pair(points[0], points[1])

    S = set()

    S.add(Point(points[0].y, points[0].x))
    S.add(Point(points[1].y, points[1].x))

    S = sorted(S)

    for i in range(2, N):
        P = points[i]
        index = bisect_left(S, Point(P.y - distance, 0))
        size_s = len(S)

        while index < size_s:
            set_point = S[index]
            Q = Point(set_point.y, set_point.x)

            if Q.x < P.x - distance:
                index = size_s - 1
                S.remove(set_point)
            if Q.y > P.y + distance:
                break

            curr_distance = get_distance(P, Q)
            if curr_distance < distance:
                distance = curr_distance
                closest = Pair(P, Q)

            index += 1

        S.append(Point(P.y, P.x))

    return closest


points = [
    Point(2, 4),
    Point(5, 3),
    Point(8, 1),
    Point(3, 6),
    Point(1, 1),
    Point(4, 4),
    Point(8, 5),
    Point(6, 2),
    Point(1, 6),
    Point(7, 4),
    Point(8, 6),
    Point(3, 2),
    Point(6, 5),
    Point(-1, 4),
]

closest_pair = get_closest_pair(points)

print(f"A: ({closest_pair.A.x}, {closest_pair.A.y})",
      f"B: ({closest_pair.B.x}, {closest_pair.B.y})")
